src/effects/RealtimeEffectStateUI.h

```diff
@@ -0,0 +1,51 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  RealtimeEffectStateUI.h
+
+  Dmitry Vedenko
+
+**********************************************************************/
+#pragma once
+
+#include "ClientData.h"
+
+class RealtimeEffectState;
+class Track;
+class EffectUIHost;
+
+class AudacityProject;
+
+//! UI state for realtime effect
+class RealtimeEffectStateUI final :
+    public ClientData::Base,
+    public ClientData::Cloneable<>
+{
+public:
+   RealtimeEffectStateUI() = default;
+   ~RealtimeEffectStateUI() override;
+
+   RealtimeEffectStateUI(const RealtimeEffectStateUI &) = default;
+   RealtimeEffectStateUI &operator=(const RealtimeEffectStateUI &) = delete;
+
+   [[nodiscard]] bool IsShown() const noexcept;
+
+   void Show(AudacityProject& project);
+   void Hide();
+   void Toggle(AudacityProject& project);
+
+   void UpdateTrackData(Track& track);
+
+   static RealtimeEffectStateUI &Get(RealtimeEffectState &state);
+   static const RealtimeEffectStateUI &Get(const RealtimeEffectState &state);
+
+   PointerType Clone() const override;
+
+private:
+   void SetActiveState(RealtimeEffectState& state); 
+   
+   class SharedState;
+   std::shared_ptr<SharedState> mSharedState;
```

Comment:
RealtimeEffectState has a `Site` which has this thing in a slot which has a `shared_ptr` to the good stuff.

I might have arranged it differently:  RealtimeEffectState has a `shared_ptr` to a SharedState structure that it declares, as a `Site`; then this would be a simple attachment and would not need to be cloneable.

Either would work.  What other attchments to states might be needed in future?  I don't know yet.  The latter way of doing it would make it impossible for them to be un-shared.

Reply:
The `shared_ptr` is only lazily initialized.
I think this means that different versions of the state in the undo history my or may not shared state.  Would that have bad consequences?

- make a new state
- do some other edit
- click the cogwheel

Then the current track list points to the shared state but the pushed undo state does not.  Does this mean I could undo, click the wheel again, and have two non-modal dialogs open at once?  Or is there something I haven't seen yet that destroys the dialog whenever you undo and redo?  (But then why have this shared state)?

But swap the last two steps, and then there is sharing.

If I am right that there could be a bug, then the shared_ptr should instead be assigned eagerly whenever a new state is made, then always copied.

So either rearrange the storage as I suggested, or used `BuildAll` in the constructor of `RealtimeEffectState` and have `RealtimeEffectStateUI` alway make the shared state.